mausezahn: Add error handling for mismatched address families

mandarg opened the pull request:
Currently, mausezahn fails silently if the user has specified IPv6
addresses using `-A` and `-B` but has forgotten the `-6` option. Add
error handling for this and the (probably uncommon) reverse case,
i.e. v4 src/dst specified, but used with the `-6` option.

Signed-off-by: Mandar Gokhale <mandarg@mandarg.com>

mandarg commented:
Testing done: Tested locally that errors were correctly thrown for both cases (IPv4 mode with at least one IPv6 address in src/dst; and IPv6 mode with at least one IPv4 address in src/dst).

I also removed an extra space before the `Invalid command line parameters!` error message because I didn't see an instance of deliberately leaving a space before – but could change it back if anyone wants.

mandarg commented:
Oh also – the motivation behind this is that I've seen people (including myself) specify src/dst IPv6 addresses but forget to specify `-6` a lot of times; and then wonder what's wrong.

tklauser commented:
Added the 'mausezahn' prefix in the commit subject (as @vkochan mentioned) and merged it as bf0df3ec1875.

Thanks a lot!

mandarg commented:
@vkochan Also, will make sure to remember this for future contributions.

vkochan commented on staging/mausezahn.c:
```diff
@@ -671,6 +671,15 @@ int getopts (int argc, char *argv[])
+			       fprintf(stderr, "You are using IPv4 mode (the default) but are providing IPv6 src/dst addresses.\n");
+			       fprintf(stderr, "Please use the -6 option for IPv6 mode.\n");
+			       return 1;
+			}
```
1. ip_dst_txt checking should be moved to the section where -B option is checking:
705                 // Set destination IP address:
 706                 if (strlen(tx.ip_dst_txt)) {  // option -B has been specified
                           ...
                           {somewhere here}
                           ...
  
2. It is better to explicitly parse ip_src_txt & ip_dst_txt addreses to check theirs AF_XXX family number because
the second case will fail:
    sudo mausezahn/mausezahn lo -6 -A evil.com -c 1 -t tcp "flags=syn"

mandarg replied:
Thanks for spotting that issue! To address 1. immediately, I have opened #167, since it's rather serious and could cause problems if anyone is building from master. (Also that check wasn't as necessary in the first place).

I'll work on moving the check to a different section and using AF_INET.

tklauser replied:
Why not just check the return values of libnet_name2addr{4,6} further below (line 689-692) instead of doing the AF_INET{,6} dance?

tklauser commented:
Reopening this and reverting bf0df3ec1875df4b167aedd0fff25e7e56408b6b for now. Sorry, I was a bit to eager to merge this and didn't properly review.

Please just push to this PR again once you have a working solution.

mandarg commented:
I'm now doing address checking using the output of the `libnet_name2addr` functions. I wrote a small test script to check for a bunch of errors and valid inputs. The script and output are at https://gist.github.com/mandarg/0f7d6d6a09f378f5d74f94757c077d4c.

I think this should be ready for review, let me know if anything else needs to be changed.

mandarg commented:
Some whitespace fixes + comment checks. Reran tests per the gist.

vkochan commented on staging/mausezahn.c:
```diff
@@ -676,11 +676,20 @@ int getopts (int argc, char *argv[])
+				  if (strncmp((char*)&tx.ip6_src,(char*)&in6addr_error,sizeof(in6addr_error))==0){
+				    fprintf(stderr, "Failed to set source IPv6 address. Please check if source is set to a valid IPv6 address.\n");
+				    return 1;
+				  }
```
I think you need to follow here indentation with a TAB instead of 2 spaces.

vkochan commented on staging/mausezahn.c:
```diff
@@ -676,11 +676,20 @@ int getopts (int argc, char *argv[])
+				    return 1;
+				  }
+				}
 			}
```
The same think regarding indentation.

vkochan commented on staging/mausezahn.c:
```diff
@@ -710,11 +719,20 @@ int getopts (int argc, char *argv[])
+					  return 1;
+					}
+			  }
 			}
```
So in the case of invalid IPv4 address we get -1 which is 255.255.255.255 IPv4 address which is OK to test for invalid host resolving, but not for IPv4 address itself I think, as 255.255.255.255 is a still valid addres to use it in the IPv4 header. The similar for IPv6, so ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff might be a valid IPv6 address. I think that it is better to check for valid IPv4/6 address format with inet_pton(...) function.

mandarg replied:
This situation seems a bit tricky to me, and I wanted to get some more input before going further.

Here is what I know so far:

* `libnet` itself defines these checks for error conditions and uses them in [sample](https://github.com/sam-github/libnet/blob/master/libnet/sample/synflood6.c#L93-L97) [code](https://github.com/sam-github/libnet/blob/master/libnet/sample/synflood6.c#L93-L97)

* The limitation where the error condition corresponds to a global broadcast address seems to be [known to some extent](https://books.google.com/books?id=lgh-akRVTuMC&pg=PA43&dq=%22is+actually+the+IP+address+255.255.255.255%22&hl=en&sa=X&ved=0ahUKEwiSztnn4YDRAhVM_IMKHeCCBfYQ6AEIGzAA#v=onepage&q=%22is%20actually%20the%20IP%20address%20255.255.255.255%22&f=false)

* As a result, it is impossible to determine whether the return value of the `libnet_name2addr` functions is a global broadcast because the user wanted to set this, or because the library is returning an error. For example, if the user is using something like `-6 -A 4.4.4.4`, the value returned by `libnet` is still going to look like a global broadcast address. I verified that this does indeed happen – so my initial statement at the top of the PR, “`mausezahn` fails silently” wasn't quite correct. It should have been “tries sending to the global broadcast address when you get invalid input.”

* However, I can't call `inet_pton` straight on the _input_ (say `tx.ip_dst_txt`) because it might be a hostname. I was depending upon `libnet` to do this for me and give me something that was guaranteed to be an IP address.

In view of this, maybe it would be good to just explicitly exclude the case where the user was putting in global broadcast addresses from validation, but do it for all other addresses. So, for example, something like `-6 -A 255.255.255.255` wouldn't throw an error, but `-6 3.3.3.3` would.

Thoughts?

vkochan replied:
So I have some thoughts, I think you can use the following alg (as a general approach):

    if (inet_pton(AF_INET, ...) == 1) {
        /* we have IPv4 address specified in X.X.X.X format */
    } else if (inet_pton(AF_INET6, ...) == 1) {
        /* we have IPv6 address in X:X::X format */
    } else if (libnet_name2addr4(...) == SUCCESS) {
        /* we have resolved IPv4 address for the hostname */
    } else ... similar

And I think it is enough.

Looking to the fiture I think if we will have:

    1) Implemented enough protos in trafgen_proto API
    2) Have generic trafgen proto generation API

then we can get rid of libnet if above's will cover all the libnet API required by mz.

mandarg replied:
I used a combination of both methods; the `libnet_name2addr` functions for all but the broadcast addresses, and `inet_pton` for covering those edge cases. This covers the additional cases pointed out by @vkochan and I have updated my [test cases](https://gist.github.com/mandarg/0f7d6d6a09f378f5d74f94757c077d4c) to include those.

I  have tried to make things as consistent as possible with the rest of the code in mausezahn.c, although as mentioned, it has inconsistencies itself in several places. (On a related note, would there be any interest in auto-formatting this using something like `astyle`? `git blame` etc. would still work as expected with `-w`.)

vkochan commented on staging/mausezahn.c:
```diff
@@ -676,11 +676,33 @@ int getopts (int argc, char *argv[])
+					((tx.ip_src = libnet_name2addr4 (l, tx.ip_src_txt, LIBNET_RESOLVE)) == -1) &&
+					// additional inet_pton() check to cover case where 255.255.255.255 is specified as source
+					(inet_pton(AF_INET, tx.ip_src_txt, &src_check) != 1)
+					)
```
Heh, this "if (" expression broken makes C look like a Lisp, a bit hard to read, may be the following would be better:

    ip_src = libnet_name2addr(...)
    if (...) {
    }

tklauser replied:
IMO you should use something like the following here:

```tx.ip_dst = = libnet_name2addr4(l, tx.ip_dst_txt, LIBNET_RESOLVE);
if (tx.ip_dst == -1) {
        /* libnet_name2addr4() returns -1 for the valid address 255.255.255.255.
         * Use an additional inet_pton() check to cover case where this address
         * is specified as destination
         */
        struct in6_addr dst_check;
        if (inet_pton(AF_INET, tx.ip_dst_txt, &dst_check) != 1) {
                fprintf(stderr, "Failed to set destination IPv4 address. Please "
                                "check if destination is set to a valid IPv4 address.\n");
                return 1;       
        }       
}
```

Similarly for the IPv6 case.

vkochan commented on staging/mausezahn.c:
```diff
@@ -710,11 +732,33 @@ int getopts (int argc, char *argv[])
-					tx.ip_dst = libnet_name2addr4 (l, tx.ip_dst_txt, LIBNET_RESOLVE);
+					struct in6_addr dst_check;
+					if (
+					(strncmp((char*)&tx.ip6_dst,(char*)&in6addr_error,sizeof(in6addr_error)) == 0) &&
```
I think you can use here libnet_in6_is_error(...) ?

vkochan commented on staging/mausezahn.c:
```diff
@@ -676,11 +676,33 @@ int getopts (int argc, char *argv[])
+					if (
+					(strncmp((char*)&tx.ip6_src,(char*)&in6addr_error,sizeof(in6addr_error)) == 0) &&
+					// additional inet_pton() check to cover case where ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff
+					// is specified as source
```
I'd use /* */ for comments which is parsed faster by the eye IMHO.

tklauser replied:
Agreed, please use /* */ comments. Also the if clause is line-wrapped a bit weird. And as @vkochan suggested, pleas use `libnet_in6_is_error()` if applicable.

tklauser commented on staging/mausezahn.c:
```diff
@@ -676,11 +676,33 @@ int getopts (int argc, char *argv[])
+					(inet_pton(AF_INET6, tx.ip_src_txt, &src_check) != 1)
+					)
+					{
+						fprintf(stderr, "Failed to set source IPv6 address. Please check if source is set to a valid IPv6 address.\n");
```
Agreed, please break this line.

mandarg commented:
Sorry for the long gap in between, but I believe this should be ready for review now.

(I don't see a good way to ACK the changes requested above; perhaps a side-effect of squashing my commit?)

tklauser commented on staging/mausezahn.c:
```diff
@@ -684,18 +684,49 @@ int getopts (int argc, char *argv[])
+						*/
+						struct in_addr src_check;
+						if (inet_pton(AF_INET, tx.ip_src_txt, &src_check) != 1){
+						fprintf(stderr, "Failed to set source"
```
This block (fprintf and return) needs one more level of indendation.

tklauser commented on staging/mausezahn.c:
```diff
@@ -718,11 +749,41 @@ int getopts (int argc, char *argv[])
+						*/
+						struct in_addr dst_check;
+						if (inet_pton(AF_INET, tx.ip_dst_txt, &dst_check) != 1){
+						fprintf(stderr, "Failed to set destination"
```
Same here. Please add one more level of indentation.

tklauser commented on staging/mausezahn.c:
```diff
@@ -684,18 +684,49 @@ int getopts (int argc, char *argv[])
+						* as source
+						*/
+						struct in6_addr src_check;
+						if (inet_pton(AF_INET6, tx.ip_src_txt, &src_check) != 1){
```
Add a space before the opening `{` please.

tklauser commented on staging/mausezahn.c:
```diff
@@ -684,18 +684,49 @@ int getopts (int argc, char *argv[])
+						* is specified as source
+						*/
+						struct in_addr src_check;
+						if (inet_pton(AF_INET, tx.ip_src_txt, &src_check) != 1){
```
Space before the opening `{` please.

tklauser commented on staging/mausezahn.c:
```diff
@@ -684,18 +684,49 @@ int getopts (int argc, char *argv[])
+						}
+					}
+				}
+				else {
```
Closing `{` and `else {` on the same line, please. I know this is not consistent across this file but while you're touching the code in question I think it's good to fix it up.

tklauser commented on staging/mausezahn.c:
```diff
@@ -684,18 +684,49 @@ int getopts (int argc, char *argv[])
+				if (strncmp((char*)&tx.ip6_src,(char*)&in6addr_error,
+					    sizeof(in6addr_error))==0)
+					printf("Failed to set source IPv6 address: %s",
+					       l->err_buf);
```
I'd keep this as it currenty is, even if it breaks the "80 characters per line" rule. IMO it's more readable that way.

tklauser commented on staging/mausezahn.c:
```diff
@@ -718,11 +749,41 @@ int getopts (int argc, char *argv[])
+						}
+					}
+				}
+				else {
```
Closing `{` and `else {` on the same line, please.

tklauser commented on staging/mausezahn.c:
```diff
@@ -718,11 +749,41 @@ int getopts (int argc, char *argv[])
+						* is specified as destination
+						*/
+						struct in_addr dst_check;
+						if (inet_pton(AF_INET, tx.ip_dst_txt, &dst_check) != 1){
```
Missing space before `{`.

mandarg commented:
Review comments from tklauser addressed.

mandarg commented:
I've addressed all the review comments, and squashed to a single commit – can this be merged, or is there something else outstanding that I missed? (if so, apologies!)

mandarg commented:
No worries, thanks!
